[TableGen] New tblgen Pattern bit to disable DAG pattern imports during GISel

jofrn opened the pull request:
Added a new DAGISelShouldIgnore property to class Instruction in Target.td, similar to FastISelShouldIgnore. This allows one to avoid a record's DAGISel .td implementation and .inc generation.

jofrn committed:
[TableGen] New tblgen Instruction bit to disable DAGISel pattern imports

Added a new DAGISelShouldIgnore property to class Instruction in Target.td, similar to FastISelShouldIgnore. This allows one to avoid a record's DAGISel .td implementation and .inc generation.

shiltian reviewed (commented):
could you add a test?

arsenm reviewed (commented):
Should add a test that a warning isn't produced on the use

jofrn committed:
[TableGen] Add test for DAGISelShouldIgnore

This test ensures no .inc code is generated in the MatcherTable for the corresponding marked .td record.

arsenm commented on llvm/test/TableGen/DAGISelEmitter-shouldignore.td:
```diff
@@ -0,0 +1,37 @@
+// RUN: llvm-tblgen --gen-dag-isel -I %p/../../include %s -DIGNORE | FileCheck %s
+// RUN: llvm-tblgen --gen-dag-isel -I %p/../../include %s --match-prefix=CHECK_LOOK | FileCheck --check-prefix=CHECK_LOOK %s
```
Should also generate globalisel and show that did import, and didn't produce the warning. The whole point is to not get warned on -warn-on-skipped-patterns

jofrn replied:
Added this check to `test/TableGen/GlobalISelEmitterSkippedPatterns.td`.

Pierre-vh commented on llvm/utils/TableGen/DAGISelEmitter.cpp:
```diff
@@ -165,8 +165,21 @@ void DAGISelEmitter::run(raw_ostream &OS) {
+    const TreePatternNode &Dst = PTM.getDstPattern();
+    if (!Dst.isLeaf()) {
+      const Record *Op = Dst.getOperator();
+      const bool shouldIgnore =
```
`CamelCase`

jofrn replied:
Thanks. [Here](https://github.com/llvm/llvm-project/pull/88382/commits/4a727fb1b7e67c4d4601afc98f29fa4c57bc4ef9#diff-6ad0fdb2565eaadd898ed286afe1367486fc0ea42880ba5cb17e8e594fccdb2d)'s the edit.

Pierre-vh commented on llvm/include/llvm/Target/Target.td:
```diff
@@ -777,6 +777,10 @@ class Instruction : InstructionEncoding {
 
+  /// Should DAGISel ignore this instruction. In cases where lowering may
+  /// be done elsewhere or is unneeded, DAGISel may skip over them.
+  bit DAGISelShouldIgnore = false;
```
I think the bit should be on the `Pat` class, not the `Instruction` class?

Do you want to disable DAG import of any pattern that uses a given instruction, or disable import for a Pat (an in/out selection pattern) ? If it's the latter then it's the wrong place for this bit.

The reason there is a FastISel bit I think is because some instructions are probably too slow to emit and FastISel needs to skip over them (?)

arsenm replied:
It probably belongs with Pat, I think this is where I got stuck last I looked at doing this

jofrn replied:
Added the new bit to `Pattern`: `bit ISelShouldIgnore = 0`. It can be set on a pattern to avoid -warn-on-skipped-patterns warnings.

jofrn committed:
[TableGen] Added tblgen ISelShouldIgnore bit to class Pattern to disable pattern imports in GISel

Via `let ISelShouldIgnore = 1`, a Pattern record can be disabled to skip pattern warnings in GlobalISel.

jofrn committed:
[TableGen] clang-format for ShouldIgnore

Pierre-vh commented on llvm/utils/TableGen/GlobalISelEmitter.cpp:
```diff
@@ -1901,7 +1901,12 @@ std::optional<CodeGenSubRegIndex *> GlobalISelEmitter::inferSubRegIndexForNode(
-Expected<RuleMatcher> GlobalISelEmitter::runOnPattern(const PatternToMatch &P) {
+std::optional<Expected<RuleMatcher>>
+GlobalISelEmitter::runOnPattern(const PatternToMatch &P) {
+  if (P.getISelShouldIgnore()) {
+    return {};
+  }
```
You can just check this in `run` before calling `runOnPattern`, then you don't need to change `runOnPattern` signature at all.

Pierre-vh commented on llvm/utils/TableGen/DAGISelEmitter.cpp:
```diff
@@ -165,8 +165,21 @@ void DAGISelEmitter::run(raw_ostream &OS) {
   Records.startTimer("Sort patterns");
   std::vector<const PatternToMatch *> Patterns;
-  for (const PatternToMatch &PTM : CGP.ptms())
+  for (const PatternToMatch &PTM : CGP.ptms()) {
```
I thought this only affected GlobalISel's import of DAG patterns, why does DAGISelEmitter need changes too?

jofrn committed:
[TableGen] style and names for GISelShouldIgnore

jofrn committed:
[TableGen] Add .inc test for GISelShouldIgnore

This test ensures the MatchTable has no patterns marked with GISelShouldIgnore.

jofrn committed:
[TableGen] Remove ShouldIgnore bit on Instruction ; not needed for dagisel

shiltian reviewed (commented):
LGTM with one nit. I'd wait to see if @arsenm and @Pierre-vh have more comments.

arsenm commented on llvm/utils/TableGen/Common/CodeGenDAGPatterns.cpp:
```diff
@@ -4378,7 +4378,8 @@ void CodeGenDAGPatterns::ParsePatterns() {
                                   InstResults, InstImpResults);
 
-    ParseOnePattern(CurPattern, Pattern, Result, InstImpResults);
+    ParseOnePattern(CurPattern, Pattern, Result, InstImpResults,
```
Can you just skip everything here if it's ignored? This currently defers it to as late as possible, and still forces you through all the rest of the machinery

jofrn replied:
It looks like this is possible. Trying it out.

arsenm replied:
Disabling both is not useful though. The only reason is to remove some globalisel migration pain

jofrn replied:
Trying to name it after its functionality, not necessarily its usage. We originally thought it to disable imports in the SelectionDAG, but it disables the imports from the perspective of GISel; so calling it with a GI prefix makes sense, but it has more general usage than that.

jofrn committed:
[TableGen] Move pattern import disabling earlier

arsenm commented on llvm/test/TableGen/GlobalISelEmitterVariadic.td:
```diff
@@ -1,4 +1,5 @@
 // RUN: llvm-tblgen -gen-global-isel -I %p/../../include -I %p/Common %s -o - | FileCheck %s
+// RUN: llvm-tblgen -gen-global-isel -I %p/../../include -I %p/Common %s -o - -DIGNORE | FileCheck %s --check-prefix=CHECK0
```
Should add -warn-on-skipped-patterns and make sure no warning is produced in stderr

jofrn replied:
These records naturally do not produce a warning with `-warn-on-skipped-patterns` with or without `GISelShouldIgnore` though. This test just checks to see if the matcher omits them.

In the other test [test/TableGen/GlobalISelEmitterSkippedPatterns.td](https://github.com/llvm/llvm-project/pull/88382/files#diff-04bf1987d3970e5d2033f31034e8cb1f4ff40ab85db1f25aad8490ec25af9895), no record is generated in the MatchTable naturally, so that one simply checks for the warning.

arsenm replied:
The entire point of the feature is to silence the warning for cases we don't want to import though

jofrn replied:
There is a check for warning omission in `GlobalISelEmitterSkippedPatterns.td`.

The .inc check test can be removed. (Alternatively, a test that skips the import/warning in global yet keeps it within the Table in dag, so that we can check both .inc and warning in the same file could be useful.) Should I just remove it or add a new test (with or without .inc check)? We have the warning check already.

jofrn commented on llvm/include/llvm/Target/TargetSelectionDAG.td:
```diff
@@ -1979,6 +1979,7 @@ class Pattern<dag patternToMatch, list<dag> resultInstrs> {
   list<dag>       ResultInstrs    = resultInstrs;
   list<Predicate> Predicates      = [];  // See class Instruction in Target.td.
   int             AddedComplexity = 0;   // See class Instruction in Target.td.
+  bit           GISelShouldIgnore = 0;
```
It is aligned around the '=', rather than shifting everyone over.

jofrn committed:
[TableGen] Remove .inc test

The intended usage of ShouldIgnore is to disable warnings. This test
checks the .inc Emitter output of GISel only. So we can remove it here.

arsenm commented on llvm/test/TableGen/GlobalISelEmitterSkippedPatterns.td:
```diff
@@ -23,6 +25,10 @@ def INSN : I<(outs GPR32:$dst), (ins GPR32:$src1, complex:$src2), []>;
+let GISelShouldIgnore = 1 in
+#endif
+// CHECK0-NOT: warning: Skipped pattern: Error: {{.*}}
 // CHECK: warning: Skipped pattern: Error: Complex suboperand x referenced by different operands: complex_rr:x:y and complex_rr:x:z.
```
Also positive check the warning in the not-DIGNORE case

arsenm replied:
I can't read

Pierre-vh commented on llvm/utils/TableGen/Common/CodeGenDAGPatterns.cpp:
```diff
@@ -4378,7 +4378,8 @@ void CodeGenDAGPatterns::ParsePatterns() {
                                   InstResults, InstImpResults);
 
-    ParseOnePattern(CurPattern, Pattern, Result, InstImpResults);
+    if (!CurPattern->getValueAsBit("GISelShouldIgnore"))
```
Wait, wouldn't that affect the DAGISel as well?
Can you add a DAGISel test to show that a GISelShouldIgnore pattern is still processed by the DAG and emitted?
Just take a basic test that checks the OPC_ opcodes emitted and add a run line with the GISelShouldIgnore bit enabled.

jofrn replied:
Yes, that would affect it in this way, which isn't what we want... Reworking back to the prior commit.

Pierre-vh commented on llvm/test/TableGen/GlobalISelEmitterSkippedPatterns.td:
```diff
@@ -23,6 +25,10 @@ def INSN : I<(outs GPR32:$dst), (ins GPR32:$src1, complex:$src2), []>;
+#ifdef IGNORE
+let GISelShouldIgnore = 1 in
+#endif
+// CHECK0-NOT: warning: Skipped pattern: Error: {{.*}}
```
nit: Can you rename this to something like `IGNORED` ? `CHECK0` doesn't tell much at first glance

Pierre-vh reviewed (changes requested):
One nit but I also have a concern with how the bit is handled

jofrn committed:
[TableGen] Move check to be within GlobalISelEmitter only.

Otherwise, the DAG is affected.

jofrn committed:
[TableGen] name change a FileCheck --check-prefix

jofrn committed:
[TableGen] Ensure DAG ISel emits records when GISelShouldIgnore'd
